### PR TITLE
Implement event mitigations

### PR DESCRIPTION
**Pull request by daeglin**
Support for player event responses (event mitigations)

### CODE REVIEW

File: frontend/src/app/game/game.service.ts [hunk 1]
```diff
@@ -109,12 +110,19 @@ export class GameService {
...
-    if (event) this.showEvent(event.title, event.text);
+    if (event) this.showEvent(event);
     if (updateChart) this.updateChart();
   }
 
-  showEvent(title: string, text: string) {
+  showEvent(event: Event) {
     this.setSpeed('pause');
-    this.eventMessages.push(`${title}: ${text}`);
+    // TODO temporary to only display events as text
```
Comment (daeglin) [inline frontend/src/app/game/game.service.ts:119]: Still a placeholder for proper event component.

File: frontend/src/app/services/event-list.ts [hunk 2]
```diff
@@ -1,33 +1,46 @@
 import {DayState} from './simulation';
-
```
Comment (daeglin) [inline frontend/src/app/services/event-list.ts:2]: Interfaces moved to events.ts

File: frontend/src/app/services/event-list.ts [hunk 3]
```diff
@@ -1,33 +1,46 @@
...
 
 // no type check below for interpolated attributes
 // TODO use rounded values
-export const eventList: Event[] = [
+export const eventTriggers: EventTrigger[] = [
+  // test events
   {
-    title: '{{stats.deaths.today}} mrtvých za jediný den',
-    text: 'Vláda podcenila situaci. Nespokojení občané žádají, tvrdší opatření. K situaci se vyjádřil předseda odborného sdružení...',
+    events: [
```
Comment (daeglin) [inline frontend/src/app/services/event-list.ts:9]: 1-n mapping of triggers to events (the code picks a random event)

File: frontend/src/app/services/events.ts [hunk 4]
```diff
@@ -1,32 +1,137 @@
...
+  condition: (stats: DayState) => boolean;
+}
+
+interface TriggerState {
+  trigger: EventTrigger;
+  timeout: number;
 }
 
 export class EventHandler {
-  triggeredEvents: TriggeredEvents[] = cloneDeep(eventList).map(e => ({...e, triggered: false}));
```
Comment (daeglin) [inline frontend/src/app/services/events.ts:10]: No need to do deep copy any more (events are copied and intepolated after the trigger is triggered).

File: frontend/src/app/services/events.ts [hunk 5]
```diff
@@ -1,32 +1,137 @@
...
+  text: string;
+  help?: string;
+  mitigations: EventMitigation[];
+}
+
+interface EventDef {
+  title: string;
+  text: string;
+  help?: string;
+  mitigations: Partial<EventMitigation>[];
```
Comment (daeglin) [inline frontend/src/app/services/events.ts:25]: Mitigations have too many parameters, we do not require all of them in the definitions.

File: frontend/src/app/services/events.ts [hunk 6]
```diff
@@ -1,32 +1,137 @@
...
 
-interface TriggeredEvents extends Event {
-  triggered: boolean;
+// infinite timeout (1 milion years)
+const infTimeout = 1_000_000 * 365;
+
+export interface EventMitigation extends MitigationEffect {
+  timeout: number;
+  label: string;
+  id?: string;
```
Comment (daeglin) [inline frontend/src/app/services/events.ts:11]: Ids are used to replace existing event mitigations.

File: frontend/src/app/services/game.ts [hunk 7]
```diff
@@ -1,17 +1,14 @@
...
-import {EventHandler} from './events';
+import {EventHandler, EventMitigation} from './events';
 import {DayState, MitigationEffect, Simulation} from './simulation';
 import {clippedLogNormalSampler, nextDay} from './utils';
 import {MitigationActions, MitigationPair, Scenario} from './scenario';
 import {Mitigations} from '../game/mitigations-control/mitigations.service';
 import {getRandomness} from './randomize';
 
-interface MitigationParams {
+export interface MitigationParams extends MitigationEffect {
```
Comment (daeglin) [inline frontend/src/app/services/game.ts:9]: There is a new interface that regularizes mitigation effects

File: frontend/src/app/services/game.ts [hunk 8]
```diff
@@ -30,7 +27,10 @@ export interface GameData {
 
 export class Game {
   readonly vaccinationStartDate = '2021-03-01';
-  readonly vaccinationPerDay = 0.01;
+  readonly vaccinationPerDay = 0.01; // TODO move to vaccination event
+  readonly infectionsWhenBordersOpen = 5;
```
Comment (daeglin) [inline frontend/src/app/services/game.ts:31]: Moved from simulation.ts

File: frontend/src/app/services/game.ts [hunk 9]
```diff
@@ -101,34 +103,53 @@ export class Game {
...
-  private calcMitigationEffect(mitigations: Mitigations, date: string): MitigationEffect {
-    let mult = 1.0;
-    let cost = 0;
-    let stabilityCost = 0;
+  addEventMitigation(mitigation: EventMitigation) {
+    this.eventHandler.addEventMitigation(mitigation);
+  }
+
+  private calcMitigationEffect(mitigations: Mitigations,
+    eventMitigations: EventMitigation[], date: string): MitigationEffect
```
Comment (daeglin) [inline frontend/src/app/services/game.ts:111]: Add event mitigations and convert to the new MitigationEffect

File: frontend/src/app/services/simulation.ts [hunk 10]
```diff
@@ -24,11 +24,11 @@ import {settings as randomizeSettings, getRandomness} from './randomize';
 import {getSeasonality, nextDay} from './utils';
 
 export interface MitigationEffect {
```
Comment (daeglin) [inline frontend/src/app/services/simulation.ts:26]: New regularized MitigationEffect interface to be used in other services

File: frontend/src/app/services/simulation.ts [hunk 11]
```diff
@@ -323,15 +318,13 @@ export class Simulation {
...
 
   private calcStats(state: SirState, modelInputs?: ModelInputs): Stats {
-    const lastStat = this.getLastStats();
-
     const detectedInfections = this.calcMetricStats('detectedInfections',
       this.getSirStateInPast(this.incubationDays).exposedNew);
     const resolvedInfections = this.calcMetricStats('resolvedInfections',
       this.getSirStateInPast(this.recoveryDuration).resistantNew + state.deathsNew);
     const deaths = this.calcMetricStats('deaths', state.deathsNew);
+    const costs = this.calcMetricStats('costs', modelInputs ? modelInputs.costToday : 0);
```
Comment (daeglin) [inline frontend/src/app/services/simulation.ts:326]: Make this look like other stats

File: frontend/src/app/services/simulation.ts [hunk 12]
```diff
@@ -74,9 +74,9 @@ export interface Stats {
   detectedInfections: MetricStats;
   resolvedInfections: MetricStats;
   deaths: MetricStats;
+  costs: MetricStats;
```
Comment (daeglin) [inline frontend/src/app/services/simulation.ts:77]: Make this look like the other stats

File: frontend/src/app/services/game.ts [hunk 13]
```diff
@@ -101,34 +103,53 @@ export class Game {
     });
   }
 
-  private calcMitigationEffect(mitigations: Mitigations, date: string): MitigationEffect {
-    let mult = 1.0;
-    let cost = 0;
-    let stabilityCost = 0;
+  addEventMitigation(mitigation: EventMitigation) {
```
Comment (daeglin) [inline frontend/src/app/services/game.ts:106]: This should be called when the player clicks one of the event mitigation options

File: frontend/src/app/services/events.ts [hunk 14]
```diff
@@ -1,32 +1,137 @@
...
+}
+
+export interface Event {
+  title: string;
+  text: string;
+  help?: string;
+  mitigations: EventMitigation[];
```
Comment (jnwltr) [inline frontend/src/app/services/events.ts:18]: We can have these optional.

**Review by jnwltr — APPROVED** [review]

**Pull request merged**